File: C/Palindrome Permutation.c
/*
Given a string, determine if a permutation of the string could form a palindrome.
Example1

Input: s = "code"
Output: False
Explanation: 
No solution
Example2

Input: s = "aab"
Output: True
Explanation: 
"aab" --> "aba"
Example3

Input: s = "carerac"
Output: True
Explanation: 
"carerac" --> "carerac"
*/


#include<stdio.h>
#include<string.h>
#include<stdlib.h>
#include<stdbool.h>
#include<assert.h>
#include<ctype.h>

bool palindrome_permutation(char s[], int stringLength)
{
    
    int frequency_count[26] = {0};
    int odd_char = 0;
    for(int i=0;i<stringLength;i++)
    {
         if((s[i] >= 'A' && s[i] <= 'Z') || (s[i] >= 'a' && s[i] <= 'z'))
        {
            char temp = tolower(s[i]);
            frequency_count[temp-'a']++;
        }
    }
    for(int i=0;i<26;i++)
    {
        if ( frequency_count[i]%2 != 0)
        {
            odd_char += 1;
        }
         if(odd_char > 1)
        {
            return false;
        }
        
    }
    return true;
}
    
int main()
{
    char Input[] = "Tacte Coa";
    int stringLength = strlen(Input);
    assert(palindrome_permutation(Input,stringLength)==0);
    
    char Input1[] = "aab";
    stringLength = strlen(Input1);
    assert(palindrome_permutation(Input1,stringLength)==1);
    
    return 0;
}

File: Meta/Practice/Python/Palindrome Permutation.py
'''
Given a string s, return true if a permutation of the string could form a 
palindrome and false otherwise.

Example 1:

Input: s = "code"
Output: false
Example 2:

Input: s = "aab"
Output: true
Example 3:

Input: s = "carerac"
Output: true
 
Constraints:

1 <= s.length <= 5000
s consists of only lowercase English letters.
'''
'''
Approach-1 :- Using set and single pass
TC :- O(n)
Sc :- O(1) for set will not grow more than 26
'''
class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        map = {}
        count = 0
        for char in s:
            map[char] = map.get(char,0)+1
            if map[char] % 2 == 0:
                count -= 1
            else:
                count += 1
        return count <= 1

'''
Approach-2 :- Using map and single pass
TC :- O(n)
SC :- O(1)
'''
class Solution:
    def canPermutePalindrome(self, s: str) -> bool:
        map = {}
        count = 0
        for char in s:
            map[char] = map.get(char,0)+1
            if map[char] % 2 == 0:
                count -= 1
            else:
                count += 1
        return count <= 1